dev_scripts/env.py

```diff
@@ -0,0 +1,527 @@
+        """Run a command for a specific container runtime.
+
+        A user's environment may have more than one container runtime [1], e.g., Podman
+        or Docker. These two runtimes have the same interface, so we can use them
+        interchangeably.
+
+        This method expects a command to run, minus the "docker" / "podman" part. Since
+        the command can be any valid command, such as "run" or "build", we can't assume
+        anything about the standard streams, so we don't affect them at all.
+
+        [1]: Technically, a container runtime is a program that implements the Container
+        Runtime Interface. We overload this term here, in lieu of a better one.
+        """
+        subprocess.run(self.runtime_cmd + list(args), check=True)
+
+    def run(
+        self, cmd, gui=DEFAULT_GUI, user=DEFAULT_USER, dry=DEFAULT_DRY, dev=DEFAULT_DEV
+    ):
+        """Run a command in a Dangerzone environment."""
+        # FIXME: Allow wiping the state of the distro before running the environment, to
+        # ensure reproducibility.
+
+        run_cmd = [
+            "run",
+            "--rm",
+            "-it",
+            "-v",
+            "/etc/localtime:/etc/localtime:ro",
+            # FIXME: Find a more secure invocation.
+            "--security-opt",
+            "seccomp=unconfined",
+            "--privileged",
+        ]
+
+        # We need to retain our UID, because we are mounting the Dangerzone source to
+        # the container.
+        if self.runtime == "podman":
+            run_cmd += ["--userns", "keep-id"]
+
+        # Compute container runtime arguments for GUI purposes.
+        if gui:
+            # Detect X11 display server connection settings.
+            env_display = os.environ.get("DISPLAY")
+            env_xauthority = os.environ.get("XAUTHORITY")
+            run_cmd += [
+                "-e",
+                f"DISPLAY={env_display}",
+                "-v",
+                "/tmp/.X11-unix:/tmp/.X11-unix:ro",
+            ]
+
+            if env_xauthority:
+                run_cmd += [
+                    "-e",
+                    f"XAUTHORITY={env_xauthority}",
+                    "-v",
+                    f"{env_xauthority}:{env_xauthority}:ro",
+                ]
+
+            # FIXME: Detect Wayland connection settings. This requires some extra
+            # settings, as we can see in this link:
+            #
+            # https://github.com/mviereck/x11docker/wiki/How-to-provide-Wayland-socket-to-docker-container
+
+        # Mount the source and the state of the distro into the container
+        dz_src = git_root()
+        dist_state = distro_state(self.distro, self.version)
+        run_cmd += [
+            "-v",
+            f"{dz_src}:/home/user/dangerzone",
+            "-v",
+            f"{dist_state}/containers:/home/user/.local/share/containers",
+            "-v",
+            f"{dist_state}/.bash_history:/home/user/.bash_history",
+        ]
```

Comment:
Ah, you interpreted keeping `.bash_history` from the same container. I actually meant mounting the history from the user's system. This way it would be shared in all containers. But I guess this is fine as well and may lead to less problems (e.g. when devs use another shell).

Reply:
Another problem with that would be that Fedora envs would see APT commands, and your host would see Dangerzone commands. Probably best to keep those separate.